openedx/openedx-k8s-harmony
# feat: add OpenSearch #21

OpenSearch integration as a Shared resource.
===


Testing instructions
---
1. Install harmony chart.
2. Install and enable k8s_harmony tutor plugin.
3. Create opensearch user.
```
tutor harmony create-opensearch-user
```

4. Verify created User from an opensearch node:
```
curl --insecure -u harmony:${HARMONY_PASSWORD} -XGET https://localhost:9200/_plugins/_security/api/internalusers/openedx-01
```
result:
```
{"openedx-01":{"hash":"","reserved":false,"hidden":false,"backend_roles":[],"attributes":{},"opendistro_security_roles":["openedx-01_role"],"static":false}}
```
5. Verify created Role from an OpenSearch node:
```
 curl --insecure -u harmony:${HARMONY_PASSWORD} -XGET https://localhost:9200/_plugins/_security/api/roles/openedx-01_role
```
result:
```
{"openedx-01_role":{"reserved":false,"hidden":false,"cluster_permissions":[],"index_permissions":[{"index_patterns":["openedx-01-uwur-*"],"fls":[],"masked_fields":[],"allowed_actions":["all"]}],"tenant_permissions":[],"static":false}}
```
6. Install openedx via tutor k8s.
8. Open CMS pod.
9. Reindex content library
```
./manage.py cms reindex_content_library
```
11. Open Studio > Create a Course > Reindex it.

Implements #21 


Issues
---
- edx-platform has several places where Elastic is accessed directly w/o `edx-search` library. This leads to AccessDenied response due to Index prefix separation.
ex.
```
File "/openedx/edx-platform/cms/djangoapps/contentstore/management/commands/reindex_course.py", line 81, in handle
    index_exists = searcher._es.indices.exists(index=index_name)  # pylint: disable=protected-access
    ....
elasticsearch.exceptions.AuthorizationException: AuthorizationException(403, '')
```
